packages/bonito-ui/src/components/form/string-list.tsx

```diff
@@ -0,0 +1,174 @@
+        [items, param, onChange]
+    );
+
+    const onItemDelete = useCallback(
+        (index: number) => {
+            const newItems = [...items];
+            newItems.splice(index, 1);
+            param.value = newItems.slice(0, newItems.length - 1) as V[K];
+            onChange?.(null, param.value);
+        },
+        [items, param, onChange]
+    );
+
+    return (
+        <Stack key={id} style={style} className={className}>
+            {items.map((item, index) => {
+                const errorMsg = validationDetails?.[index];
+                return (
+                    <StringListItem
+                        key={index}
+                        index={index}
+                        value={item}
+                        label={param.label}
+                        errorMsg={errorMsg}
+                        placeholder={param.placeholder}
+                        onChange={onItemChange}
+                        onDelete={onItemDelete}
+                        disableDelete={index === items.length - 1}
+                    ></StringListItem>
+                );
+            })}
+        </Stack>
+    );
+}
+
+interface StringListItemProps {
+    index: number;
+    value: string;
+    label?: string;
+    onChange: (index: number, value: string) => void;
+    onDelete: (index: number) => void;
+    placeholder?: string;
+    disableDelete?: boolean;
+    errorMsg?: string;
+}
+
+function StringListItem(props: StringListItemProps) {
+    const {
+        index,
+        value,
+        label,
+        onChange,
+        onDelete,
+        disableDelete,
+        errorMsg,
+        placeholder,
+    } = props;
+    const styles = useStringListItemStyles(props);
+    const ariaLabel = `${label || ""} ${index + 1}`;
+    return (
+        <Stack
+            key={index}
+            horizontal
+            verticalAlign="center"
+            styles={styles.container}
+        >
+            <Stack.Item grow={1}>
+                <TextField
+                    styles={styles.input}
+                    value={value}
+                    ariaLabel={ariaLabel}
+                    placeholder={placeholder}
+                    errorMessage={errorMsg}
+                    onChange={(_, newValue) => {
+                        onChange(index, newValue || "");
+                    }}
+                />
+            </Stack.Item>
+            <IconButton
+                styles={styles.button}
+                iconProps={{ iconName: "Delete" }}
+                ariaLabel={`${translate("bonito.ui.form.delete")} ${ariaLabel}`}
+                onClick={() => {
+                    onDelete(index);
+                }}
+                disabled={disableDelete}
+            />
+        </Stack>
+    );
+}
+
+function useStringListItemStyles(props: StringListItemProps) {
+    const theme = useAppTheme();
+    const { disableDelete } = props;
+
+    const itemPadding = {
+        padding: "11px 8px 11px 12px",
+    };
+
+    return {
```

Comment:
I see your point about overriding styles, though in general it makes me a bit nervous to do that as specific inner styles tend to be implementation details which change between releases.

Can we at least memoize this function? These nested style objects are being created once for every textbox in the list per keystroke right now regardless of whether anything changed.

Reply:
Actually using `styles` props to customize component style is what they recommended. For example for `TextField`, there is a [interface](https://developer.microsoft.com/en-us/fluentui#/controls/web/textfield#ITextFieldStyles) to show what part of a `TextField` you can customize. I think it will be stable in Major versions. Fluent V9 is another story, they don't recommend to use `mergeStyles` any more, see: https://react.fluentui.dev/?path=/docs/concepts-migration-from-v8-components-theme-migration--page#mergestyles---mergeclasses

Reply:
Agreed that we should memorize this function

Reply:
Jeez, Fluent UI 8 -> 9 is going to be a ton of fun (whenever the portal supports 9). I'm fine with keeping it and just adding some memoization.